src/main/java/org/folio/service/transactions/batch/BatchTransactionService.java

```diff
@@ -0,0 +1,229 @@
+      Set<BatchTransactionServiceInterface> batchTransactionStrategies) {
+    this.dbClientFactory = dbClientFactory;
+    this.transactionDAO = transactionDAO;
+    this.fundService = fundService;
+    this.budgetService = budgetService;
+    this.ledgerService = ledgerService;
+    serviceMap = new EnumMap<>(TransactionType.class);
+    batchTransactionStrategies.forEach(
+      strategy -> serviceMap.put(strategy.getTransactionType(), strategy));
+  }
+
+  public Future<Void> processBatch(Batch batch, RequestContext requestContext) {
+    try {
+      BatchTransactionChecks.sanityChecks(batch);
+    } catch (Exception ex) {
+      logger.error("Sanity checks before processing batch transactions failed, batch={}", Json.encode(batch), ex);
+      return Future.failedFuture(ex);
+    }
+    populateMetadata(batch, requestContext.getHeaders());
+    DBClient client = dbClientFactory.getDbClient(requestContext);
+    BatchTransactionHolder holder = new BatchTransactionHolder(transactionDAO, fundService, budgetService, ledgerService);
+    return client.withTrans(conn -> holder.setup(batch, conn)
+      .map(v -> {
+        BatchTransactionChecks.checkBudgetsAreActive(holder);
+        prepareCreatingTransactions(holder);
+        prepareUpdatingTransactions(holder);
+        preparePatchingTransactions(holder);
+        BatchTransactionChecks.checkRestrictedBudgets(holder);
+        return null;
+      })
+      .compose(v -> applyChanges(holder, conn))
+    ).onSuccess(v -> logger.info("All batch transaction operations were successful."))
+      .onFailure(t -> logger.error("Error when batch processing transactions, batch={}",
+        Json.encode(batch), t));
+  }
+
+  private void prepareCreatingTransactions(BatchTransactionHolder holder) {
+    if (holder.getAllTransactionsToCreate().isEmpty()) {
+      return;
+    }
+    try {
+      int totalPrepared = 0;
+      for (TransactionType trType : transactionTypesInOrder) {
+        List<Transaction> transactionsForType = getByType(trType, holder.getAllTransactionsToCreate());
+        if (!transactionsForType.isEmpty()) {
+          getBatchServiceForType(trType).prepareCreatingTransactions(transactionsForType, holder);
+          totalPrepared += transactionsForType.size();
+        }
+      }
+      logger.info("Successfully prepared {} transactions for creation", totalPrepared);
+    } catch (Exception ex) {
+      List<String> ids = holder.getAllTransactionsToCreate().stream().map(Transaction::getId).toList();
+      logger.error("Failed to prepare transactions for creation, ids = {}", ids, ex);
+      throw ex;
+    }
+  }
+
+  private void prepareUpdatingTransactions(BatchTransactionHolder holder) {
+    if (holder.getAllTransactionsToUpdate().isEmpty()) {
+      return;
+    }
+    try {
+      // NOTE: holder.getAllTransactionsToUpdate() can change after processing payments/pending payments,
+      // so we don't group transactions by type before processing them.
```

Comment:
for reviewers: initially there was grouping transactions before preparing them (by action and type), which made the code simpler and more stream-oriented. The problem with this approach is that it creates a copy of the list of transactions to update which does not get updated after preparing transactions to create. This comment to remind developers not to group transactions, so we don't risk missing the updates to holder.getAllTransactionsToUpdate().